clients/src/main/java/org/apache/kafka/clients/admin/ListConsumerGroupsResult.java

```diff
@@ -34,70 +32,77 @@
+                    ArrayList<ConsumerGroupListing> curValid = new ArrayList<>();
+                    for (Object object : objects) {
+                        if (object instanceof Throwable) {
+                            curErrors.add((Throwable) object);
+                        } else {
+                            curValid.add((ConsumerGroupListing) object);
+                        }
+                    }
+                    if (!curErrors.isEmpty()) {
+                        all.completeExceptionally(curErrors.get(0));
+                    } else {
+                        all.complete(curValid);
+                    }
+                    valid.complete(curValid);
+                    errors.complete(curErrors);
                 }
             }
+        });
+    }
 
-            KafkaFutureImpl<ConsumerGroupListing> future = new KafkaFutureImpl<>();
-            future.complete(innerIter.next());
-            return future;
-        }
+    /**
+     * Returns a future that yields either an exception, or the full set of consumer group
+     * listings.
+     *
+     * In the event of a failure, the future yields nothing but the first exception which
+     * occurred.
+     */
+    public KafkaFutureImpl<Collection<ConsumerGroupListing>> all() {
+        return all;
     }
 
     /**
-     * Return an iterator of futures for ConsumerGroupListing objects; the returned future will throw exception
-     * if we cannot get a complete collection of consumer listings.
+     * Returns a future which yields just the valid listings.
+     *
+     * This future never fails with an error, no matter what happens.  Errors are completely
+     * ignored.  If nothing can be fetched, an empty collection is yielded.
+     * If there is an error, but some results can be returned, this future will yield
+     * those partial results.  When using this future, it is a good idea to also check
+     * the errors future so that errors can be displayed and handled.
      */
-    public Iterator<KafkaFuture<ConsumerGroupListing>> iterator() {
-        return new FutureConsumerGroupListingIterator();
+    public KafkaFutureImpl<Collection<ConsumerGroupListing>> valid() {
+        return valid;
     }
 
     /**
-     * Return a future which yields a full collection of ConsumerGroupListing objects; will throw exception
-     * if we cannot get a complete collection of consumer listings.
+     * Returns a future which yields just the errors which occurred.
+     *
+     * If this future yields a non-empty collection, it is very likely that elements are
+     * missing from the valid() set.
+     *
+     * This future itself never fails with an error.  In the event of an error, this future
+     * will successfully yield a collection containing at least one exception.
      */
-    public KafkaFuture<Collection<ConsumerGroupListing>> all() {
-        return flattenFuture;
+    public KafkaFutureImpl<Collection<Throwable>> errors() {
```

Comment:
This is meta: I'm wondering if return a collection of errors would help, since users do not have any information on which nodes returns each error. For example, how would user learn more with two UnknownTopicOrPartition errors compared with just getting one? My feeling is it may create more confusion than more information it brings.

Reply:
Users do have information on which nodes returned errors, though: the exception text always mentions a node name.

Some code may choose to only look at the exception code, which is also a valid choice (in general some exception codes should be retried, others not.)  But printing the exception text, which most code will also do, will help clear up what is going on in the cluster.